File: client/src/services/KeepsService.js
import { AppState } from "../AppState.js"
import { Keep } from "../models/Keep.js"
import { logger } from "../utils/Logger.js"
import { api } from "./AxiosService.js"

class KeepsService {

    async getKeeps() {
        const res = await api.get('api/keeps')
        const newKeeps = res.data.map((keep) => new Keep(keep))
        AppState.keeps = newKeeps
    }

    async setActiveKeep(keepId) {
        AppState.activeKeep = {}
        const res = await api.get(`api/keeps/${keepId}`)
        const keepData = new Keep(res.data)
        AppState.activeKeep = keepData
    }
    async createKeep(keepData) {
        const res = await api.post('api/keeps', keepData)
        AppState.keeps.push(new Keep(res.data))
    }

    async destroyKeep(keepId) {
        const res = await api.delete(`api/keeps/${keepId}`)
        AppState.keeps = AppState.keeps.filter((keep) => keep.Id != keepId)
        AppState.activeKeep = null
    }


}
export const keepsService = new KeepsService()

File: server/Services/KeepsService.cs




namespace Keeper.Services
{
    public class KeepsService
    {

        private readonly KeepsRepository _keepsRepository;

        public KeepsService(KeepsRepository keepsRepository)
        {
            _keepsRepository = keepsRepository;
        }

        internal Keep CreateKeep(Keep keepData)
        {
            Keep keep = _keepsRepository.CreateKeep(keepData);
            return keep;
        }

        internal string DestroyKeep(int keepId, string userId)
        {
            Keep keep = GetKeepById(keepId);

            if (keep.CreatorId != userId)
            {
                throw new Exception("Not your Keep to Delete!");
            }
            _keepsRepository.DestroyKeep(keepId);
            return $"{keep.Name} has been Deleted!";
        }

        internal Keep GetKeepById(int keepId)
        {
            Keep keep = _keepsRepository.GetKeepById(keepId);

            if (keep == null)
            {
                throw new Exception($"{keepId} is invalid!");
            }
            return keep;

        }
        internal List<Keep> GetKeeps()
        {
            List<Keep> keeps = _keepsRepository.GetKeeps();
            return keeps;
        }


        internal List<Keep> GetKeepsByProfileId(string profileId, string userId)
        {

            List<Keep> keeps = _keepsRepository.GetKeepsByProfileId(profileId, userId);
            return keeps;
        }

        internal Keep UpdateKeep(int keepId, string userId, Keep keepData)
        {
            Keep ogKeep = GetKeepById(keepId);

            if (ogKeep.CreatorId != userId)
            {
                throw new Exception("Not your Keep to Update");
            }
            ogKeep.Name = keepData.Name ?? ogKeep.Name;
            ogKeep.Description = keepData.Description ?? ogKeep.Description;

            Keep keep = _keepsRepository.UpdateKeep(ogKeep);
            return keep;
        }
    }

}
